Bug(githubactions-golang):  Main Builder failed with error `go.mod file not found`

## Describe the Bug

/plugin `githubactions-golang`

GitHub Actions log:

```sh
go: go.mod file not found in current directory or any parent directory; see 'go help modules'
[5](https://github.com/daniel-hutao/dtm-test-go/runs/6139165057?check_suite_focus=true#step:2:5)
Error: Process completed with exit code 1.
```

## Screenshots

![image](https://user-images.githubusercontent.com/18692628/164888079-7f6c5136-e872-4a37-890d-20fab5c483e2.png)

/kind bug
/assign @daniel-hutao

--- a/examples/gitops.yaml
+++ b/examples/gitops.yaml
@@ -39,7 +39,7 @@ tools:
         profile: "-race -covermode=atomic"
         output: "coverage.out"
     docker:
-      enable: False
+      enable: True
       repo: ${{github-repo-scaffolding-golang.default.outputs.repo}}
 - name: argocd
   instanceID: default

--- a/examples/gitops.yaml
+++ b/examples/gitops.yaml
@@ -40,7 +40,10 @@ tools:
         output: "coverage.out"
     docker:
       enable: True
-      repo: ${{github-repo-scaffolding-golang.default.outputs.repo}}
+      registry:
+        type: dockerhub
+        username: [[ dockerhubUsername ]]
+        repository: ${{github-repo-scaffolding-golang.default.outputs.repo}}
 - name: argocd
   instanceID: default
   options:

--- a/internal/pkg/plugin/githubactions/golang/create.go
+++ b/internal/pkg/plugin/githubactions/golang/create.go
@@ -41,10 +41,19 @@ func Create(options map[string]interface{}) (map[string]interface{}, error) {
 
 	// if docker is enabled, create repo secrets for DOCKERHUB_USERNAME and DOCKERHUB_TOKEN
 	if opts.Docker != nil && opts.Docker.Enable {
-		if err := ghClient.AddRepoSecret("DOCKERHUB_USERNAME", viper.GetString("dockerhub_username")); err != nil {
+		dockerhubUsername := viper.GetString("dockerhub_username")
+		dockerhubToken := viper.GetString("dockerhub_token")
+		if dockerhubUsername == "" {
+			return nil, fmt.Errorf("DockerHub Username is empty")
+		}
+		if dockerhubToken == "" {
+			return nil, fmt.Errorf("DockerHub Token is empty")
+		}
+
+		if err := ghClient.AddRepoSecret("DOCKERHUB_USERNAME", dockerhubUsername); err != nil {
 			return nil, err
 		}
-		if err := ghClient.AddRepoSecret("DOCKERHUB_TOKEN", viper.GetString("dockerhub_token")); err != nil {
+		if err := ghClient.AddRepoSecret("DOCKERHUB_TOKEN", dockerhubToken); err != nil {
 			return nil, err
 		}
 	}

--- a/internal/pkg/plugin/githubactions/golang/create.go
+++ b/internal/pkg/plugin/githubactions/golang/create.go
@@ -41,18 +41,11 @@ func Create(options map[string]interface{}) (map[string]interface{}, error) {
 
 	// if docker is enabled, create repo secrets for DOCKERHUB_USERNAME and DOCKERHUB_TOKEN
 	if opts.Docker != nil && opts.Docker.Enable {
-		dockerhubUsername := viper.GetString("dockerhub_username")
 		dockerhubToken := viper.GetString("dockerhub_token")
-		if dockerhubUsername == "" {
-			return nil, fmt.Errorf("DockerHub Username is empty")
-		}
 		if dockerhubToken == "" {
 			return nil, fmt.Errorf("DockerHub Token is empty")
 		}
 
-		if err := ghClient.AddRepoSecret("DOCKERHUB_USERNAME", dockerhubUsername); err != nil {
-			return nil, err
-		}
 		if err := ghClient.AddRepoSecret("DOCKERHUB_TOKEN", dockerhubToken); err != nil {
 			return nil, err
 		}

--- a/internal/pkg/plugin/githubactions/golang/jobs.go
+++ b/internal/pkg/plugin/githubactions/golang/jobs.go
@@ -1,5 +1,55 @@
 package golang
 
+import "fmt"
+
+type RegistryType string
+
+const (
+	RegistryDockerhub RegistryType = "dockerhub"
+	RegistryHarbor    RegistryType = "harbor"
+)
+
+// Build is the struct for githubacions job.
+type Build struct {
+	Enable  bool
+	Command string
+}
+
+// Test is the struct for githubacions job.
+type Test struct {
+	Enable   bool
+	Command  string
+	Coverage Coverage
+}
+
+// Docker is the struct for githubacions job.
+type Docker struct {
+	Enable   bool
+	Registry Registry
+}
+
+type Registry struct {
+	// only support dockerhub now
+	Type       RegistryType
+	Username   string
+	Repository string
+}
+
+// Coverage is the struct for githubacions job.
+type Coverage struct {
+	Enable  bool
+	Profile string
+	Output  string
+}
+
+// Tag is the struct for githubacions job.
+type Tag struct {
+}
+
+// Image is the struct for githubacions job.
+type Image struct {
+}
+
 func (b *Build) Validate() []error {
 	retErrors := make([]error, 0)
 
@@ -19,41 +69,22 @@ func (t *Test) Validate() []error {
 func (d *Docker) Validate() []error {
 	retErrors := make([]error, 0)
 
-	// TODO(daniel-hutao): what should we validate here?
+	if !d.Enable {
+		return retErrors
+	}
+
+	if d.Registry.Username == "" {
+		retErrors = append(retErrors, fmt.Errorf("registry username is empty"))
+	}
+
+	// default to dockerhub
+	if d.Registry.Type == "" {
+		d.Registry.Type = RegistryDockerhub
+		return retErrors
+	}
+	if d.Registry.Type != RegistryDockerhub && d.Registry.Type != RegistryHarbor {
+		retErrors = append(retErrors, fmt.Errorf("registry type != (dockerhub || harbor)"))
+	}
 
 	return retErrors
-}
-
-// Build is the struct for githubacions job.
-type Build struct {
-	Enable  bool
-	Command string
-}
-
-// Test is the struct for githubacions job.
-type Test struct {
-	Enable   bool
-	Command  string
-	Coverage Coverage
-}
-
-// Docker is the struct for githubacions job.
-type Docker struct {
-	Enable bool
-	Repo   string
-}
-
-// Coverage is the struct for githubacions job.
-type Coverage struct {
-	Enable  bool
-	Profile string
-	Output  string
-}
-
-// Tag is the struct for githubacions job.
-type Tag struct {
-}
-
-// Image is the struct for githubacions job.
-type Image struct {
 }

--- a/internal/pkg/plugin/githubactions/golang/render.go
+++ b/internal/pkg/plugin/githubactions/golang/render.go
@@ -19,10 +19,6 @@ func renderTemplate(workflow *github.Workflow, options *Options) (string, error)
 	if opts.Build == nil {
 		opts.Build = &Build{false, ""}
 	}
-	// let Docker is false when empty
-	if opts.Docker == nil {
-		opts.Docker = &Docker{false, ""}
-	}
 
 	//if use default {{.}}, it will confict (github actions vars also use them)
 	t, err := template.New("githubactions").Delims("[[", "]]").Parse(workflow.WorkflowContent)

--- a/internal/pkg/plugin/githubactions/golang/templates.go
+++ b/internal/pkg/plugin/githubactions/golang/templates.go
@@ -9,6 +9,11 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
+    - uses: actions/checkout@v2
+    - name: Set up Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.17
     - name: Build
       run: [[- if not .Build.Command]] go build ./...[[- else]] [[.Build.Command]][[- end]]
   [[- else]]
@@ -82,6 +87,11 @@ jobs:
   build:
     runs-on: ubuntu-latest
     steps:
+    - uses: actions/checkout@v2
+    - name: Set up Go
+      uses: actions/setup-go@v2
+      with:
+        go-version: 1.17
     - name: Build
       run: [[- if not .Build.Command]] go build ./...[[- else]] [[.Build.Command]][[- end]]
   [[- else]]

--- a/internal/pkg/plugin/githubactions/golang/templates.go
+++ b/internal/pkg/plugin/githubactions/golang/templates.go
@@ -67,14 +67,14 @@ jobs:
       - name: Login to Docker Hub
         uses: docker/login-action@v1
         with:
-          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          username: [[.Docker.Registry.Username]]
           password: ${{ secrets.DOCKERHUB_TOKEN }}
       - name: Build and push
         id: docker_build
         uses: docker/build-push-action@v2
         with:
           push: true
-          tags: ${{ secrets.DOCKERHUB_USERNAME }}/[[- if not .Docker.Repo]][[.Repo]][[- else]][[.Docker.Repo]][[- end]]:${{needs.tag.outputs.new_tag}}
+          tags: [[.Docker.Registry.Username]]/[[- if not .Docker.Registry.Repository]][[.Repo]][[- else]][[.Docker.Registry.Repository]][[- end]]:${{needs.tag.outputs.new_tag}}
   [[- end]]
 `
 
@@ -152,13 +152,13 @@ jobs:
       - name: Login to Docker Hub
         uses: docker/login-action@v1
         with:
-          username: ${{ secrets.DOCKERHUB_USERNAME }}
+          username: [[.Docker.Registry.Username]]
           password: ${{ secrets.DOCKERHUB_TOKEN }}
       - name: Build and push
         id: docker_build
         uses: docker/build-push-action@v2
         with:
           push: true
-          tags: ${{ secrets.DOCKERHUB_USERNAME }}/[[- if not .Docker.Repo]][[.Repo]][[- else]][[.Docker.Repo]][[- end]]:${{needs.tag.outputs.new_tag}}
+          tags: [[.Docker.Registry.Username]]/[[- if not .Docker.Registry.Repository]][[.Repo]][[- else]][[.Docker.Registry.Repository]][[- end]]:${{needs.tag.outputs.new_tag}}
   [[- end]]
 `

--- a/internal/pkg/pluginengine/change.go
+++ b/internal/pkg/pluginengine/change.go
@@ -70,8 +70,8 @@ func getChanges(smgr statemanager.Manager, cfg *configloader.Config, commandType
 	}
 
 	log.Debugf("Changes for the plan:")
-	for _, c := range changes {
-		log.Debugf(c.String())
+	for i, c := range changes {
+		log.Debugf("Change - %d/%d -> %s", i+1, len(changes), c.String())
 	}
 
 	return changes, nil
@@ -86,7 +86,7 @@ func execute(smgr statemanager.Manager, changes []*Change) map[string]error {
 
 	for i, c := range changes {
 		log.Separatorf("Processing progress: %d/%d.", i+1, numOfChanges)
-		log.Infof("Processing: %s (%s) -> %s ...", c.Tool.Name, c.Tool.InstanceID, c.ActionName)
+		log.Infof("Processing: (%s/%s) -> %s ...", c.Tool.Name, c.Tool.InstanceID, c.ActionName)
 
 		var succeeded bool
 		var err error
@@ -101,7 +101,7 @@ func execute(smgr statemanager.Manager, changes []*Change) map[string]error {
 			for _, e := range errs {
 				log.Errorf("Error: %s.", e)
 			}
-			log.Errorf("The outputs reference in tool %s (%s) can't be resolved. Please double check your config.", c.Tool.Name, c.Tool.InstanceID)
+			log.Errorf("The outputs reference in tool (%s/%s) can't be resolved. Please double check your config.", c.Tool.Name, c.Tool.InstanceID)
 
 			// not executing this change since its input isn't valid
 			continue
@@ -144,9 +144,9 @@ func execute(smgr statemanager.Manager, changes []*Change) map[string]error {
 
 // handleResult is used to Write the latest StatesMap to the Backend.
 func handleResult(smgr statemanager.Manager, change *Change) error {
-	log.Debugf("Start: \n%s", string(smgr.GetStatesMap().Format()))
+	log.Debugf("Start -> StatesMap now is:\n%s", string(smgr.GetStatesMap().Format()))
 	defer func() {
-		log.Debugf("End:\n%s", string(smgr.GetStatesMap().Format()))
+		log.Debugf("End -> StatesMap now is:\n%s", string(smgr.GetStatesMap().Format()))
 	}()
 
 	if !change.Result.Succeeded {
@@ -162,7 +162,7 @@ func handleResult(smgr statemanager.Manager, change *Change) error {
 			log.Debugf("Failed to delete state %s: %s.", key, err)
 			return err
 		}
-		log.Successf("Tool %s (%s) delete done.", change.Tool.Name, change.Tool.InstanceID)
+		log.Successf("Tool (%s/%s) delete done.", change.Tool.Name, change.Tool.InstanceID)
 		return nil
 	}
 
@@ -179,6 +179,6 @@ func handleResult(smgr statemanager.Manager, change *Change) error {
 		log.Debugf("Failed to add state %s: %s.", key, err)
 		return err
 	}
-	log.Successf("Tool %s(%s) %s done.", change.Tool.Name, change.Tool.InstanceID, change.ActionName)
+	log.Successf("Tool (%s/%s) %s done.", change.Tool.Name, change.Tool.InstanceID, change.ActionName)
 	return nil
 }

--- a/internal/pkg/pluginengine/change_helper.go
+++ b/internal/pkg/pluginengine/change_helper.go
@@ -79,7 +79,7 @@ func changesForApply(smgr statemanager.Manager, cfg *configloader.Config) ([]*Ch
 
 			if state == nil {
 				// tool not in the state, create, no need to Read resource before Create
-				description := fmt.Sprintf("Tool %s (%s) found in config but doesn't exist in the state, will be created.", tool.Name, tool.InstanceID)
+				description := fmt.Sprintf("Tool (%s/%s) found in config but doesn't exist in the state, will be created.", tool.Name, tool.InstanceID)
 				changes = append(changes, generateCreateAction(&tool, description))
 			} else {
 				// tool found in the state
@@ -90,7 +90,7 @@ func changesForApply(smgr statemanager.Manager, cfg *configloader.Config) ([]*Ch
 
 				if drifted(tool.Options, state.Options) {
 					// tool's config differs from State's, Update
-					description := fmt.Sprintf("Tool %s (%s) config drifted from the state, will be updated.", tool.Name, tool.InstanceID)
+					description := fmt.Sprintf("Tool (%s/%s) config drifted from the state, will be updated.", tool.Name, tool.InstanceID)
 					changes = append(changes, generateUpdateAction(&tool, description))
 				} else {
 					// tool's config is the same as State's
@@ -103,15 +103,15 @@ func changesForApply(smgr statemanager.Manager, cfg *configloader.Config) ([]*Ch
 
 					if resource == nil {
 						// tool exists in the state, but resource doesn't exist, Create
-						description := fmt.Sprintf("Tool %s (%s) state found but it seems the tool isn't created, will be created.", tool.Name, tool.InstanceID)
+						description := fmt.Sprintf("Tool (%s/%s) state found but it seems the tool isn't created, will be created.", tool.Name, tool.InstanceID)
 						changes = append(changes, generateCreateAction(&tool, description))
 					} else if drifted(resource, state.Resource) {
 						// resource drifted from state, Update
-						description := fmt.Sprintf("Tool %s (%s) drifted from the state, will be updated.", tool.Name, tool.InstanceID)
+						description := fmt.Sprintf("Tool (%s/%s) drifted from the state, will be updated.", tool.Name, tool.InstanceID)
 						changes = append(changes, generateUpdateAction(&tool, description))
 					} else {
 						// resource is the same as the state, do nothing
-						log.Debugf("Tool %s (%s) is the same as the state, do nothing.", tool.Name, tool.InstanceID)
+						log.Debugf("Tool (%s/%s) is the same as the state, do nothing.", tool.Name, tool.InstanceID)
 					}
 				}
 			}
@@ -152,7 +152,7 @@ func changesForDelete(smgr statemanager.Manager, cfg *configloader.Config, isFor
 				}
 			}
 
-			description := fmt.Sprintf("Tool %s (%s) will be deleted.", tool.Name, tool.InstanceID)
+			description := fmt.Sprintf("Tool (%s/%s) will be deleted.", tool.Name, tool.InstanceID)
 			changes = append(changes, generateDeleteAction(&tool, description))
 		}
 	}
@@ -185,7 +185,7 @@ func GetChangesForDestroy(smgr statemanager.Manager) ([]*Change, error) {
 	for i := len(batchesOfTools) - 1; i >= 0; i-- {
 		batch := batchesOfTools[i]
 		for _, tool := range batch {
-			description := fmt.Sprintf("Tool %s (%s) will be deleted.", tool.Name, tool.InstanceID)
+			description := fmt.Sprintf("Tool (%s/%s) will be deleted.", tool.Name, tool.InstanceID)
 			changes = append(changes, generateDeleteAction(&tool, description))
 		}
 	}

--- a/internal/pkg/pluginengine/cmd_apply.go
+++ b/internal/pkg/pluginengine/cmd_apply.go
@@ -53,7 +53,7 @@ func Apply(configFile, varFile string, continueDirectly bool) error {
 	errsMap := execute(smgr, changes)
 	if len(errsMap) != 0 {
 		for k, e := range errsMap {
-			log.Errorf("Error: key(%s) -> msg: %s", k, e)
+			log.Errorf("Errors Map: key(%s) -> value(%s)", k, e)
 		}
 		return errors.New("some error(s) occurred during plugins apply process")
 	}

--- a/internal/pkg/pluginengine/topological_sort.go
+++ b/internal/pkg/pluginengine/topological_sort.go
@@ -64,7 +64,7 @@ func topologicalSort(tools []configloader.Tool) ([][]configloader.Tool, error) {
 				}
 			}
 		}
-		log.Debugf("BATCH: %s", batch)
+		log.Debugf("BATCH: %v", batch)
 
 		// there are still nodes unprocessed but there is no node whose dependency is solved
 		// this means there might be a loop in the graph

--- a/internal/pkg/show/config/plugin/githubactions-golang.go
+++ b/internal/pkg/show/config/plugin/githubactions-golang.go
@@ -41,5 +41,10 @@ var GithubActionsGolangDefaultConfig = `tools:
     docker:
       # docker build/push related, default to false
       enable: True
-      # dockerhub image repo; please change the value below.
-      repo: YOUR_DOCKERHUB_REPOSITORY`
+      registry:
+        # dockerhub or harbor, default to dockerhub
+        type: dockerhub
+        # dockerhub/harbor username
+        username: YOUR_DOCKERHUB_USERNAME
+        # dockerhub/harbor image repository name
+        repository: YOUR_DOCKERHUB_REPOSITORY`

--- a/internal/pkg/statemanager/manager.go
+++ b/internal/pkg/statemanager/manager.go
@@ -111,7 +111,7 @@ func (m *manager) DeleteState(key StateKey) error {
 func (m *manager) GetOutputs(key StateKey) (interface{}, error) {
 	state := m.GetState(key)
 	if state == nil {
-		return nil, fmt.Errorf("key %s not in state", key)
+		return nil, fmt.Errorf("key (%s) not in state, it may be failed when \"Create\"", key)
 	}
 
 	if value, ok := state.Resource["outputs"]; ok {

--- a/test/e2e/yaml/e2e-config.yaml
+++ b/test/e2e/yaml/e2e-config.yaml
@@ -24,6 +24,9 @@ tools:
         enable: True
     docker:
       enable: True
+      registry:
+        type: dockerhub
+        username: dtme2etest
 - name: argocd
   instanceID: default
   options:

